### PR TITLE
Set up SEARCH_PATH only once when running an update command

### LINKED ISSUE #5316: Running update-sql with PostgreSQL sets the search path multiple times
### Search first

- [X] I searched and no similar issues were found

### Description

When using PostgreSQL and the CLI, running `update-sql` will create a script that sets the search path way too many times.
For example, it is set once before the first update to the `databasechangeloglock` table, and four times between that and the next relevant statement.

This is an issue with Liquibase version 4.25.0, but it was a problem before that as well.

Here's the changelog I used:
```
<?xml version="1.0" encoding="UTF-8"?>
<databaseChangeLog
  xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-latest.xsd
                      http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd
                      http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-latest.xsd">

  <changeSet id="1" author="Liquibase">
    <createTable tableName="test_table">
      <column name="test_id" type="int">
        <constraints primaryKey="true" />
      </column>
      <column name="test_column" type="varchar" />
    </createTable>
  </changeSet>

  <changeSet id="2" author="Liquibase">
    <createTable tableName="test_table_2">
      <column name="test_id" type="int">
        <constraints primaryKey="true" />
      </column>
      <column name="test_column" type="varchar" />
    </createTable>
  </changeSet>

  <changeSet id="3" author="Liquibase">
    <createTable tableName="test_table_3">
      <column name="test_id" type="int">
        <constraints primaryKey="true" />
      </column>
      <column name="test_column" type="varchar" />
    </createTable>
  </changeSet>

</databaseChangeLog>
```

And here's the generated sql script:
```
SET SEARCH_PATH TO public, "$user","public";

-- Lock Database
UPDATE public.databasechangeloglock SET LOCKED = TRUE, LOCKEDBY = 'GXCN4031 (192.168.0.111)', LOCKGRANTED = NOW() WHERE ID = 1 AND LOCKED = FALSE;

SET SEARCH_PATH TO public, "$user","public";

SET SEARCH_PATH TO public, "$user","public";

-- *********************************************************************
-- Update Database Script
-- *********************************************************************
-- Change Log: changelog.xml
-- Ran at: 05/12/2023, 12:45
-- Against: lbadminuser@jdbc:postgresql://localhost:54322/lbadmindb
-- Liquibase version: 4.25.0
-- *********************************************************************

SET SEARCH_PATH TO public, "$user","public";

-- Changeset changelog.xml::1::Liquibase
SET SEARCH_PATH TO public, "$user","public";

CREATE TABLE public.test_table (test_id INTEGER NOT NULL, test_column VARCHAR, CONSTRAINT test_table_pkey PRIMARY KEY (test_id));

INSERT INTO public.databasechangelog (ID, AUTHOR, FILENAME, DATEEXECUTED, ORDEREXECUTED, MD5SUM, DESCRIPTION, COMMENTS, EXECTYPE, CONTEXTS, LABELS, LIQUIBASE, DEPLOYMENT_ID) VALUES ('1', 'Liquibase', 'changelog.xml', NOW(), 1, '9:c31cbcc82ef990951e46cc173776542c', 'createTable tableName=test_table', '', 'EXECUTED', NULL, NULL, '4.25.0', '1791137079');

-- Changeset changelog.xml::2::Liquibase
SET SEARCH_PATH TO public, "$user","public";

CREATE TABLE public.test_table_2 (test_id INTEGER NOT NULL, test_column VARCHAR, CONSTRAINT test_table_2_pkey PRIMARY KEY (test_id));

INSERT INTO public.databasechangelog (ID, AUTHOR, FILENAME, DATEEXECUTED, ORDEREXECUTED, MD5SUM, DESCRIPTION, COMMENTS, EXECTYPE, CONTEXTS, LABELS, LIQUIBASE, DEPLOYMENT_ID) VALUES ('2', 'Liquibase', 'changelog.xml', NOW(), 2, '9:a5a63d62e7fcd7d78d1acaa6c255fd1a', 'createTable tableName=test_table_2', '', 'EXECUTED', NULL, NULL, '4.25.0', '1791137079');

-- Changeset changelog.xml::3::Liquibase
SET SEARCH_PATH TO public, "$user","public";

CREATE TABLE public.test_table_3 (test_id INTEGER NOT NULL, test_column VARCHAR, CONSTRAINT test_table_3_pkey PRIMARY KEY (test_id));

INSERT INTO public.databasechangelog (ID, AUTHOR, FILENAME, DATEEXECUTED, ORDEREXECUTED, MD5SUM, DESCRIPTION, COMMENTS, EXECTYPE, CONTEXTS, LABELS, LIQUIBASE, DEPLOYMENT_ID) VALUES ('3', 'Liquibase', 'changelog.xml', NOW(), 3, '9:b09d5fc1dd4fc3177bd8e6426ba6d962', 'createTable tableName=test_table_3', '', 'EXECUTED', NULL, NULL, '4.25.0', '1791137079');

-- Release Database Lock
SET SEARCH_PATH TO public, "$user","public";

UPDATE public.databasechangeloglock SET LOCKED = FALSE, LOCKEDBY = NULL, LOCKGRANTED = NULL WHERE ID = 1;

SET SEARCH_PATH TO public, "$user","public";
```

### Steps To Reproduce

1. Set up Liquibase with a PostgreSQL database connection
2. Create a changelog - add multiple changesets for a clearer example
3. Run `update-sql`

### Expected/Desired Behavior

The search path should be set only once

### Liquibase Version

4.25.0

### Database Vendor & Version

PostgreSQL - all versions

### Liquibase Integration

CLI

### Liquibase Extensions

_No response_

### OS and/or Infrastructure Type/Provider

_No response_

### Additional Context

It's **not** an issue with other databases like MSSQL or DB2

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

### PR DESCRIPTION
**Pull request by MalloD12**
## Impact

- [x] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 

## Description

Fixes #5316.

This PR provides a new `DatabaseUtils` method to set up `SEARCH_PATH` variable permanently when running any update family command. Previously when running a `update-sql` command on a Postgres DB we could see multiple `SET SEARCH_PATH...` statements in the generated SQL output. It seems that was happening because of the `DatabaseUtils.initializeDatabase()` call performed in the `rollback()` method from `PostgresDatabase` implementation. With this change that's not happening anymore, but SQL output will be a bit different now, since what we will see is:
```
ALTER DATABASE lbcat SET SEARCH_PATH TO test_schema_daniel, "$user","public";   
```
only once.

## Things to be aware of
- This a Postgres specific change.

## Things to worry about
- I guess we might need to apply the same logic in PRO (`update-one-changeset`) if we are happy with these changes.

### CODE REVIEW

**Comment by MalloD12** [comment]
This is still a work in progress since I need to add some tests to this PR, but I would appreciate it if you guys could start reviewing these changes in the meantime.

From a testing point of view what I have done so far is to execute `update`/`update-sql` commands using a changelog with a few changesets and having:
```
<changeSet author="lbuser" id="4-table" labels="testtable4" contextFilter="none" failOnError="false">
        <createTable tableName="testtable4">
            <column name="supplier_id" type="numeric(4)"/>
            <column name="supplier_name" type="TTTTVARCHAR(50)"/>
        </createTable>
    </changeSet>
```
as one of them. Which will cause the rollback method to be triggered. 

- Run `update`/`update-sql`commands without proving any specific schema. Output was:
```
-- Create Database Lock Table
CREATE TABLE public.databasechangeloglock (ID INTEGER NOT NULL, LOCKED BOOLEAN NOT NULL, LOCKGRANTED TIMESTAMP WITHOUT TIME ZONE, LOCKEDBY VARCHAR(255), CONSTRAINT databasechangeloglock_pkey PRIMARY KEY (ID));

-- Initialize Database Lock Table
DELETE FROM public.databasechangeloglock;

INSERT INTO public.databasechangeloglock (ID, LOCKED) VALUES (1, FALSE);

-- Lock Database
UPDATE public.databasechangeloglock SET LOCKED = TRUE, LOCKEDBY = 'Daniels-MacBook-Pro.local (192.168.1.28)', LOCKGRANTED = NOW() WHERE ID = 1 AND LOCKED = FALSE;

-- Create Database Change Log Table
CREATE TABLE public.databasechangelog (ID VARCHAR(255) NOT NULL, AUTHOR VARCHAR(255) NOT NULL, FILENAME VARCHAR(255) NOT NULL, DATEEXECUTED TIMESTAMP WITHOUT TIME ZONE NOT NULL, ORDEREXECUTED INTEGER NOT NULL, EXECTYPE VARCHAR(10) NOT NULL, MD5SUM VARCHAR(35), DESCRIPTION VARCHAR(255), COMMENTS VARCHAR(255), TAG VARCHAR(255), LIQUIBASE VARCHAR(20), CONTEXTS VARCHAR(255), LABELS VARCHAR(255), DEPLOYMENT_ID VARCHAR(10));

-- *********************************************************************
-- Update Database Script
-- *********************************************************************
-- Change Log: changelog/xml/showSummaryWithLabels.xml
-- Ran at: 5/01/24, 3:54 PM
-- Against: postgres@jdbc:postgresql://localhost:5438/lbcat
-- Liquibase version: [local build]
-- *********************************************************************

ALTER DATABASE lbcat SET SEARCH_PATH TO public, "$user","public";

-- Changeset changelog/xml/showSummaryWithLabels.xml::1-table::lbuser
CREATE TABLE public.testtable1 (supplier_id numeric(4), supplier_name VARCHAR(50));

....

-- Release Database Lock
UPDATE public.databasechangeloglock SET LOCKED = FALSE, LOCKEDBY = NULL, LOCKGRANTED = NULL WHERE ID = 1;

```
Also `update` was successfully executed storing changes in the `public` schema, as expected.
  
- Run `update`/`update-sql`commands proving `--default-schema-name=test_schema_daniel` schema. Output was:

```
-- Create Database Lock Table
CREATE TABLE test_schema_daniel.databasechangeloglock (ID INTEGER NOT NULL, LOCKED BOOLEAN NOT NULL, LOCKGRANTED TIMESTAMP WITHOUT TIME ZONE, LOCKEDBY VARCHAR(255), CONSTRAINT databasechangeloglock_pkey PRIMARY KEY (ID));

-- Initialize Database Lock Table
DELETE FROM test_schema_daniel.databasechangeloglock;

INSERT INTO test_schema_daniel.databasechangeloglock (ID, LOCKED) VALUES (1, FALSE);

ALTER DATABASE lbcat SET SEARCH_PATH TO test_schema_daniel, "$user","public";

-- Changeset changelog/xml/showSummaryWithLabels.xml::1-table::lbuser
CREATE TABLE test_schema_daniel.testtable1 (supplier_id numeric(4), supplier_name VARCHAR(50));

-- Changeset changelog/xml/showSummaryWithLabels.xml::4-table::lbuser
CREATE TABLE test_schema_daniel.testtable4 (supplier_id numeric(4), supplier_name TTTTVARCHAR(50));

...

Process finished with exit code 0
```

- Also execute same test (only for Postgres) that Enzo comment [here](https://github.com/liquibase/liquibase/pull/2812#issuecomment-1126380309).

**NOTE**: just in case I would like to mention part of the provided SQL outputs have been deleted to not make comments too long because of the length of them.

File: liquibase-standard/src/main/java/liquibase/database/core/DatabaseUtils.java [hunk 1]
```diff
@@ -75,12 +75,53 @@ public static void initializeDatabase(String defaultCatalogName, String defaultS
...
+    public static void initializePostgresSearchPathPermanently(String defaultCatalogName, String defaultSchemaName, Database database) {
+
+        final Executor executor = Scope.getCurrentScope().getSingleton(ExecutorService.class).getExecutor("jdbc", database);
+
+        if (database instanceof PostgresDatabase && defaultSchemaName != null) {
+            String searchPath = null;
+            try {
+                searchPath = executor.queryForObject(new RawSqlStatement("SHOW SEARCH_PATH"), String.class);
+            } catch (Throwable e) {
+                Scope.getCurrentScope().getLog(DatabaseUtils.class).info("Cannot get search_path", e);
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/database/core/DatabaseUtils.java:95]: Maybe this should be logged at a different level than info?

File: liquibase-standard/src/main/java/liquibase/database/core/DatabaseUtils.java [hunk 2]
```diff
@@ -75,12 +75,53 @@ public static void initializeDatabase(String defaultCatalogName, String defaultS
...
+                                return obj;
+                            }
+                            return ((PostgresDatabase) database).quoteObject(obj, Schema.class);
+                        });
+                    }
+
+                    try {
+                        executor.execute(new RawSqlStatement(String.format("ALTER DATABASE %s SET SEARCH_PATH TO %s", database.getLiquibaseCatalogName(), finalSearchPath)));
+                    } catch (Throwable e) {
+                        Scope.getCurrentScope().getLog(DatabaseUtils.class).info("Cannot set search_path at database level", e);
```
Comment (StevenMassaro) [inline liquibase-standard/src/main/java/liquibase/database/core/DatabaseUtils.java:120]: Log level?

**Review by StevenMassaro — COMMENTED** [review]
Looks good from here

File: liquibase-standard/src/main/java/liquibase/database/core/DatabaseUtils.java (hunk 1 above)
Comment (MalloD12) [reply liquibase-standard/src/main/java/liquibase/database/core/DatabaseUtils.java:95]: Right, thinking of updating this as a warning and the other one you comment below as an error. Would that be the correct logging level for both of these messages?

Thanks,
Daniel.

Comment (StevenMassaro) [reply liquibase-standard/src/main/java/liquibase/database/core/DatabaseUtils.java:95]: Sounds right to me

**Pull request closed without merging**